Semantic segmentation support + semantic augmentations

SundarRajan98 opened the pull request.

SundarRajan98 committed:
Adding initial changes for semantic segmentation support

SundarRajan98 committed:
Fixing build issues and adding python example

SundarRajan98 committed:
Adding random mask pixel notebook

SundarRajan98 committed:
Merge branch 'develop' into semantic_segm_support

SundarRajan28 committed:
Merge branch 'develop' into semantic_segm_support

SundarRajan28 committed:
Add python example for semantic augmentations

SundarRajan28 commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +450,268 @@ MasterGraph::reset() {
+            }
+            prev_label = in_mask_buffer[i];
+        }
+        auto dist = std::uniform_int_distribution<int64_t>(0, unique_labels.size() - 1);
```
[Blocking] `unique_labels` can be empty when a sample only has background pixels, so constructing std::uniform_int_distribution(0, unique_labels.size()-1) underflows to (0,-1) and is undefined. In practice this trips assertions before you even know the image was empty. Please guard the case where the label set is empty and either skip the sample or emit a sentinel bbox instead of calling the distribution.

SundarRajan28 commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -1099,26 +1380,167 @@ TensorListVector* MasterGraph::create_coco_meta_data_reader(const char *source_p
+#pragma omp parallel for num_threads(_user_batch_size)
+        for (unsigned i = 0; i < _user_batch_size; i++) {
+            auto rng = rngs[i];
+            std::vector<unsigned long> dims{input->at(i)->dims().at(0), 1};
```
[Blocking] `dims` is hard-coded to `{input->dims(0), 1}` here, so the second coordinate we report back is permanently 0. Any consumer of `rocalRandomMaskPixel` therefore never gets a valid X coordinate and the API is unusable. Please use the actual width (dims().at(1)) instead of the literal 1, and make the same fix in the other two branches below.

SundarRajan28 commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -1099,26 +1380,167 @@ TensorListVector* MasterGraph::create_coco_meta_data_reader(const char *source_p
+                    } else {
+                        start.push_back(id++);
+                        foreground_count.push_back(count++);
+                        while (mask_buffer[id] > _random_mask_pixel_value) {
```
[Blocking] The inner `while (mask_buffer[id] > ...)` loop never checks that `id < buffer_size`, so when the last foreground run extends to the end of the tensor we read one element past the allocation (e.g. mask_buffer[buffer_size]). Please break or guard on `id < buffer_size` here and in the equality variant below before dereferencing.

SundarRajan28 commented on rocAL/include/meta_data/meta_data.h:
```diff
@@ -194,6 +197,25 @@ struct PolygonMask : public BoundingBox {
+        _info.img_id = img_id;
+    }
+    std::vector<int>& get_pixelwise_label() override { return _pixelwise_label; }
+    void set_pixelwise_label(std::vector<int>& pixelwise_label) override { _pixelwise_label = std::move(pixelwise_label); }
```
[Blocking] `PixelwiseMask::set_pixelwise_label` is marked `override`, but `MetaData` never declares a setter. This fails to compile because there is no virtual you can override. Please add a matching virtual setter to `MetaData` (just like `get_pixelwise_label`) or drop the override and keep the method on the derived class only.

SundarRajan28 commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -244,13 +417,42 @@ void COCOMetaDataReader::read_all(const std::string &path) {
+                                    parser.SkipValue();
+                                }
+                            }
+                            if (!rle_str.empty()) {
```
[Blocking] The two `counts` branches are swapped: when COCO stores counts as a string we immediately call `rleInit` with an empty `rle_uints`, and when counts is an array we pass an empty string to `rleFrString`. Crowd masks therefore never get decoded. Please call `rleFrString` when `counts` was a string and `rleInit` when you actually collected the integer array.

SundarRajan28 commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -71,11 +71,15 @@ RocalMetaData
-    if (mask) {
-        return context->master_graph->create_coco_meta_data_reader(source_path, is_output, MetaDataReaderType::COCO_META_DATA_READER, MetaDataType::PolygonMask, ltrb, is_box_encoder, avoid_class_remapping, aspect_ratio_grouping, is_box_iou_matcher);
+    if (is_polygon_mask) {
+        return context->master_graph->create_coco_meta_data_reader(source_path, is_output, MetaDataReaderType::COCO_META_DATA_READER, MetaDataType::PolygonMask, ltrb, is_box_encoder, avoid_class_remapping, aspect_ratio_grouping);
```
[Blocking] When `is_polygon_mask` or `is_pixelwise_mask` is true we now drop the `is_box_iou_matcher` flag instead of forwarding it, so users can no longer enable IoU matching for those readers. Please thread the `is_box_iou_matcher` argument through in both mask branches, just like the bounding-box branch below.

SundarRajan28 commented on rocAL_pybind/rocal_pybind.cpp:
```diff
@@ -1000,6 +1006,84 @@ py::class_<rocalListOfTensorList>(m, "rocalListOfTensorList")
+            int cnt = 0;
+            py::dict mask_select_polygon_dict;
+            for (int j = 0; j < mask_ids.size(); j++) {
+                py::list select_mask_polygon_list;
```
[Blocking] `sel_vertices_counts[i]` only contains entries for masks that were actually present, but this loop iterates over every requested `mask_id` and blindly indexes `sel_vertices_count_per_image[j]`. As soon as a requested id is missing (very common) we segfault. Please guard the access (e.g. search for the mask id or check the vector length) instead of assuming every requested id exists.

SundarRajan28 commented on rocAL/include/pipeline/seed_rng.h:
```diff
@@ -0,0 +1,44 @@
+#include <random>
+#include <bits/stdc++.h>
```
[Non-blocking] Including `<bits/stdc++.h>` from a public header makes the build non-portable (it only exists on libstdc++) and drags in a huge amount of unrelated code into every TU that includes `master_graph.h`. Please include just the headers you actually use (`<random>`, `<vector>`, …) instead.

SundarRajan28 replied:
**Blocking (portability/perf):** Please avoid '#include <bits/stdc++.h>' in a public header (non-standard GCC-only, large compile-time hit). Also consider adding the standard AMD license header here and keeping '#pragma once' at the top.

SundarRajan28 committed:
Fix build issue

SundarRajan28 committed:
Initial fix

SundarRajan28 committed:
More fixes

SundarRajan28 committed:
Add rocAL support for Set 1 augmentations: GaussianNoise, ShotNoise, Spatter, Log, ColorJitter, Water

- Add node headers and implementations for GaussianNoise, ShotNoise, Spatter, Log, ColorJitter, Water
- Add C API declarations and implementations in rocal_api_augmentation.h/cpp
- Add Python bindings in rocal_pybind.cpp and fn.py

SundarRajan28 committed:
Add rocAL support for Set 2 augmentations: ChannelPermute, ColorToGreyscale, JpegCompressionDistortion, Lut, Posterize, Solarize

- Add node headers and implementations for ChannelPermute, ColorToGreyscale, JpegCompressionDistortion, Lut, Posterize, Solarize
- Add C API declarations and implementations in rocal_api_augmentation.h/cpp
- Add Python bindings in rocal_pybind.cpp and fn.py

SundarRajan28 committed:
Add rocAL support for Set 3 augmentations: TensorSum, TensorMin, TensorMax, TensorMean, TensorStdDev

- Add node header and implementation for TensorStatistics (supporting Sum, Min, Max, Mean, StdDev operations)
- Add C API declarations and implementations in rocal_api_augmentation.h/cpp
- Add Python bindings in rocal_pybind.cpp and fn.py

SundarRajan28 committed:
Add VX_EXT_RPP version check for Set 1 augmentation APIs

SundarRajan28 committed:
Add VX_EXT_RPP version check for Set 2 augmentation APIs

SundarRajan28 committed:
Add VX_EXT_RPP version check for Set 3 augmentation APIs

SundarRajan28 committed:
Update vx_rpp version checks to 3.1.5 for Set 1 augmentations

- Update VX_EXT_RPP_CHECK_VERSION to (3, 1, 5) in API file
- Update THROW messages to indicate vx_rpp >= 3.1.5
- Add version checks in node files: node_gaussian_noise.cpp,
  node_shot_noise.cpp, node_spatter.cpp, node_water.cpp,
  node_color_jitter.cpp, node_log.cpp

SundarRajan28 committed:
Update vx_rpp version checks to 3.1.6 for Set 2 augmentations

- Update VX_EXT_RPP_CHECK_VERSION to (3, 1, 6) in API file
- Update THROW messages to indicate vx_rpp >= 3.1.6
- Add version checks in node files: node_channel_permute.cpp,
  node_color_to_greyscale.cpp, node_lut.cpp, node_posterize.cpp,
  node_solarize.cpp, node_jpeg_compression_distortion.cpp

SundarRajan28 committed:
Update vx_rpp version checks to 3.1.7 for Set 3 augmentations

- Update VX_EXT_RPP_CHECK_VERSION to (3, 1, 7) in API file
- Update THROW messages to indicate vx_rpp >= 3.1.7
- Add version checks in node_tensor_statistics.cpp for TensorSum,
  TensorMin, TensorMax, TensorMean, and TensorStdDev operations

SundarRajan28 committed:
Add C++ and Python tests for Set 1 augmentations

C++ tests (unit_tests.cpp):
- Case 64: rocalGaussianNoise (random)
- Case 65: rocalGaussianNoiseFixed
- Case 66: rocalShotNoise (random)
- Case 67: rocalShotNoiseFixed
- Case 68: rocalSpatter
- Case 69: rocalLog
- Case 70: rocalColorJitter (random)
- Case 71: rocalColorJitterFixed
- Case 72: rocalWater (random)
- Case 73: rocalWaterFixed

Python tests (unit_test.py):
- gaussian_noise, shot_noise, spatter, water, color_jitter

Shell scripts updated:
- testAllScripts.sh: Added C++ test entries
- unit_tests.sh: Added Python test entries

SundarRajan28 committed:
Add C++ and Python tests for Set 2 augmentations

Test cases 74-83 for:
- ChannelPermute (74)
- JpegCompressionDistortion (75, 76-fixed)
- LUT (77)
- Posterize (78, 79-fixed)
- Solarize (80, 81-fixed)
- Log (82)
- ColorToGreyscale (83)

SundarRajan28 committed:
Add C++ and Python tests for Set 3 augmentations

Test case 84 for tensor reduction operations:
- TensorSum
- TensorMin
- TensorMax
- TensorMean
- TensorStdDev

SundarRajan28 committed:
Update rocAL APIs and TensorStdDevNode

SundarRajan28 committed:
Add API changes

SundarRajan28 committed:
Update test case number

SundarRajan28 committed:
Fix issue with Log

SundarRajan28 committed:
Update unit tests

SundarRajan28 committed:
Unit test changes

SundarRajan28 committed:
Unit tests fix

SundarRajan28 committed:
Spatter API changes

SundarRajan28 committed:
Update test case numbers for Set 2 augmentations

Shift test case numbers by +1 (75-83) to follow after Set 1
which now ends at case 74 due to added Spatter random variant.

SundarRajan28 committed:
Update test case number for Set 3 tensor reduction

Change test case from 83 to 84 to follow after Set 2
which now ends at case 83 (ColorToGreyscale).

SundarRajan28 committed:
Unit test changes

SundarRajan28 committed:
Update copyright year to 2026

SundarRajan98 committed:
Merge branch 'develop' into new_augs_set1

SundarRajan28 committed:
Update copyright year to 2026

SundarRajan28 committed:
Update copyright year to 2026

SundarRajan28 committed:
Update vx_rpp version check

SundarRajan28 committed:
Move VX_EXT_RPP_CHECK_VERSION to beginning of API functions

Move version check to the very beginning of each function,
before any variable assignments or context validation, for
all Set 1 augmentation APIs.

SundarRajan28 committed:
Move VX_EXT_RPP_CHECK_VERSION to beginning of API functions

SundarRajan28 committed:
Move VX_EXT_RPP_CHECK_VERSION to beginning of API functions

SundarRajan28 committed:
Add support for per-image RLE masks

SundarRajan28 committed:
Merge remote-tracking branch 'origin/develop' into semantic_segm_support

SundarRajan28 committed:
Fix unit test issues

SundarRajan28 committed:
Fix python tests and notebook

SundarRajan28 committed:
Update API docs

SundarRajan28 committed:
Merge remote-tracking branch 'ToT/develop' into new_augs_set1

SundarRajan28 committed:
Merge branch 'new_augs_set1' into new_augs_set2

SundarRajan28 committed:
Merge branch 'new_augs_set2' into new_augs_set3

SundarRajan28 committed:
Merge remote-tracking branch 'sr_dev/new_augs_set3' into semantic_segm_support

SundarRajan28 commented on thirdparty/coco_RLE/include/maskApi.h:
```diff
@@ -0,0 +1,63 @@
+/**************************************************************************
+ * Microsoft COCO Toolbox.      version 2.0
+ * Data, paper, and tutorials available at:  http://mscoco.org/
+ * Code written by Piotr Dollar and Tsung-Yi Lin, 2015.
```
**Blocking (licensing/compliance):** This header references "Licensed under the Simplified BSD License [see coco/license.txt]", but that license file is not present in this repo. Please vendor the actual license text (and ideally a link/commit to the upstream source) under 'thirdparty/coco_RLE/' (e.g. add 'thirdparty/coco_RLE/LICENSE') so downstream redistribution is unambiguous.

SundarRajan28 commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -244,13 +447,37 @@ void COCOMetaDataReader::read_all(const std::string &path) {
                         }
-                    } else if ((_output->get_metadata_type() == MetaDataType::PolygonMask) && 0 == std::strcmp(internal_key, "segmentation")) {
-                        if (parser.PeekType() == kObjectType) {
+                    } else if ((_output->get_metadata_type() == MetaDataType::PolygonMask || _output->get_metadata_type() == MetaDataType::PixelwiseMask) && 0 == std::strcmp(internal_key, "segmentation")) {
```
**Blocking (correctness):** COCO 'segmentation' can be an object (RLE) (commonly when 'iscrowd == 1'). For 'MetaDataType::PolygonMask' this code path currently asserts 'kArrayType' and can crash on valid COCO JSON. Suggest: if 'segmentation' is an object, call 'parser.SkipValue()' for PolygonMask (or handle it), and/or gate segmentation parsing on 'iscrowd == 0' before asserting on the type.

SundarRajan28 commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -244,13 +447,37 @@ void COCOMetaDataReader::read_all(const std::string &path) {
+                                    parser.NextArrayValue();
+                                    w = parser.GetInt();
+                                    parser.NextArrayValue();
+                                } else if (0 == std::strcmp(another_key, "counts")) {
```
**Blocking (security/robustness):** RLE 'counts' are parsed as 'int' and stored into 'uint32_t' without validation. Negative or oversized values can wrap and later overflow internal buffers in 'rleMerge' / pixelwise decode (potential OOB write/crash on malformed input). Please validate: counts are non-negative, total pixels match 'h*w', and 'h/w' match the image size before constructing/merging RLEs.

SundarRajan28 commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -137,8 +149,195 @@ void COCOMetaDataReader::print_map_contents() {
+            RLE M;
+            rleInit(&M, 0, 0, 0, 0);
+            if (!mask.counts_str.empty()) {
+                rleFrString(&M, const_cast<char *>(mask.counts_str.c_str()), mask.h, mask.w);
```
**Blocking (const-correctness/robustness):** Calling 'rleFrString(&M, const_cast<char*>(mask.counts_str.c_str()), ...)' relies on the helper not mutating the input buffer. Prefer passing a mutable 'std::string' buffer (e.g. 'std::string s = mask.counts_str; rleFrString(&M, s.data(), ...)') or updating the helper signature to accept 'const char*'. Also please validate 'mask.h/mask.w' against 'img_size' before merging.

SundarRajan28 commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -1099,26 +1587,169 @@ TensorListVector* MasterGraph::create_coco_meta_data_reader(const char *source_p
+    output_random_mask_pixel.clear();
+    output_random_mask_pixel.resize(_user_batch_size * 2);
+    if (_is_random_mask_pixel_foreground == false) {
+#pragma omp parallel for num_threads(_user_batch_size)
```
**Blocking (performance/reproducibility):** '#pragma omp parallel for num_threads(_user_batch_size)' can massively oversubscribe the CPU (batch size != thread count). Also 'SeededRNG' is re-created and seeded from 'std::random_device' per call, ignoring 'rocalSetSeed()' / 'ParameterFactory' seed, making results non-reproducible. Recommend: use '_cpu_num_threads' (or min), and derive RNG state from rocAL's existing seeded RNG so runs are deterministic when a seed is set.

SundarRajan28 commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +450,475 @@ MasterGraph::reset() {
+
+TensorList *MasterGraph::get_random_object_bbox(rocalTensorList *input, RandomObjectBBoxFormat format,
+                                                int k_largest, float foreground_prob, bool cache_objects) {
+    SeededRNG<std::mt19937, 4> rngs(_user_batch_size);
```
**Blocking (reproducibility/perf):** 'SeededRNG<std::mt19937, 4> rngs(_user_batch_size)' is constructed per-call inside 'get_random_object_bbox()', reseeding from 'std::random_device'. This makes outputs change across calls even with a fixed rocAL seed and adds overhead. Consider making RNG state a 'MasterGraph' member seeded from rocAL's seed (or accepting an explicit seed parameter).

SundarRajan28 commented on rocAL_pybind/rocal_pybind.cpp:
```diff
@@ -1037,6 +1043,97 @@ py::class_<rocalListOfTensorList>(m, "rocalListOfTensorList")
+    });
+    m.def("getRandomObjectBBox", [](RocalContext context, RocalRandomObjectBBoxFormat format,
+                                     int k_largest, float foreground_prob, bool cache_objects) {
+        rocalTensorList *boxes = RocalRandomObjectBBox(context, format, k_largest, foreground_prob, cache_objects);
```
**Non-blocking (Python API safety):** 'getRandomObjectBBox' returns 'py::array' views that alias internal C++ buffers. This matches other rocAL APIs, but users may keep arrays past the next call and hit stale/overwritten data. If feasible, consider returning an owning 'py::array_t<unsigned>' copy (or clearly documenting the lifetime/validity window).

SundarRajan28 commented on rocAL/include/api/rocal_api_meta_data.h:
```diff
@@ -317,6 +326,49 @@ extern "C" void ROCAL_API_CALL rocalBoxIouMatcher(RocalContext p_context, std::v
+/*! \brief Pick a random pixel coordinate from PixelwiseMask metadata
+ * \ingroup group_rocal_meta_data
+ * \param [in] p_context rocAL context
+ * \return RocalTensorList of 2D coordinates (x, y) per image (valid only for PixelwiseMask metadata)
```
**Non-blocking (API docs):** The docstring says random pixel coords are '(x, y)', but the implementation/tests treat the output as '(row, col)' / '(y, x)'. Please align the docs (and ideally name the order explicitly).

SundarRajan28 commented on rocAL/include/meta_data/coco_meta_data_reader.h:
```diff
@@ -21,12 +21,15 @@ THE SOFTWARE.
 #include "meta_data/meta_data.h"
 #include "meta_data/meta_data_reader.h"
 #include "pipeline/timing_debug.h"
+#include "maskApi.h"
```
**Non-blocking (maintainability):** 'maskApi.h' is included from this public header, but it looks only needed in the '.cpp'. Consider moving it to 'rocAL/source/meta_data/coco_meta_data_reader.cpp' to avoid leaking global typedefs ('uint', 'siz', 'RLE') to downstream includes.

SundarRajan28 commented on rocAL/CMakeLists.txt:
```diff
@@ -240,6 +240,7 @@ if(${BUILD_ROCAL})
                 ${CMAKE_CURRENT_BINARY_DIR}
                 include
                 include/api
+                ./../thirdparty/coco_RLE/include/
```
**Non-blocking (build hygiene):** Adding '../thirdparty/coco_RLE/include/' via global 'include_directories()' makes 'maskApi.h' visible everywhere. If possible, prefer target-scoped includes (e.g. 'target_include_directories(rocal PRIVATE ...)') so the third-party header doesn't leak into consumers/builds that don't need it.

fiona-gladwin replied:
Check this comment

SundarRajan28 commented on rocAL/include/pipeline/master_graph.h:
```diff
@@ -81,6 +82,36 @@ const __m256i avx_pkdMaskB = _mm256_setr_epi32(0x80808002, 0x80808005, 0x8080800
+};
+
+// Simple hash for caching
+inline size_t fast_hash_buffer(const void* data, size_t size) {
```
**Non-blocking (correctness):** The cache key uses a simple FNV-1a hash over the mask buffer. Hash collisions are possible and would return wrong cached boxes. If 'cache_objects' is used for correctness-sensitive paths, consider storing additional discriminators (dims + a secondary hash) or validating on hit.

fiona-gladwin commented on rocAL/include/api/rocal_api_meta_data.h:
```diff
@@ -78,15 +78,17 @@ extern "C" RocalMetaData ROCAL_API_CALL rocalCreateTFReaderDetection(RocalContex
  * \return RocalMetaData object, can be used to inquire about the rocal's output (processed) tensors
  */
-extern "C" RocalMetaData ROCAL_API_CALL rocalCreateCOCOReader(RocalContext rocal_context, const char* source_path, bool is_output, bool mask = false, bool ltrb = true, bool is_box_encoder = false, bool avoid_class_remapping = false, bool aspect_ratio_grouping = false, bool is_box_iou_matcher = false);
+extern "C" RocalMetaData ROCAL_API_CALL rocalCreateCOCOReader(RocalContext rocal_context, const char* source_path, bool is_output, bool is_polygon_mask = false, bool is_pixelwise_mask = false, bool ltrb = true, bool is_box_encoder = false, bool avoid_class_remapping = false, bool aspect_ratio_grouping = false, bool is_box_iou_matcher = false);
```
Require changes in external API, check all the docs the examples where this API is used

fiona-gladwin commented on rocAL/include/api/rocal_api_meta_data.h:
```diff
@@ -317,6 +326,49 @@ extern "C" void ROCAL_API_CALL rocalBoxIouMatcher(RocalContext p_context, std::v
+ * \param [in] cache_objects if true, cache object bounding boxes for repeated inputs
+ * \return RocalTensorList of 4-element bounding boxes per image (valid only for PixelwiseMask metadata)
+ */
+extern "C" RocalTensorList ROCAL_API_CALL RocalRandomObjectBBox(RocalContext p_context, RocalRandomObjectBBoxFormat format,
```
rocal, keep it small letters similar to other API

fiona-gladwin commented on rocAL/include/api/rocal_api_types.h:
```diff
@@ -405,6 +405,20 @@ enum RocalExternalSourceMode {
+ *
+ * Output format for RocalRandomObjectBBox.
+ */
+enum RocalRandomObjectBBoxFormat {
```
Please add equivalent enum in the commons.h and it needs to be used everywhere internally, required for serialization

fiona-gladwin commented on rocAL/include/meta_data/coco_meta_data_reader.h:
```diff
@@ -38,24 +41,45 @@ class COCOMetaDataReader : public MetaDataReader {
+        int mask_idx = -1;
+        int h = -1;
+        int w = -1;
+        std::string counts_str;
+        std::vector<uint32_t> counts;
```
Add comment for these variables

fiona-gladwin commented on rocAL/include/meta_data/coco_meta_data_reader.h:
```diff
@@ -50,12 +52,23 @@ class COCOMetaDataReader : public MetaDataReader {
+    {52, 47}, {53, 48}, {54, 49}, {55, 50}, {56, 51}, {57, 52}, {58, 53}, {59, 54}, {60, 55}, {61, 56},  \
+    {62, 57}, {63, 58}, {64, 59}, {65, 60}, {67, 61}, {70, 62}, {72, 63}, {73, 64}, {74, 65}, {75, 66},  \
+    {76, 67}, {77, 68}, {78, 69}, {79, 70}, {80, 71}, {81, 72}, {82, 73}, {84, 74}, {85, 75}, {86, 76},  \
+    {87, 77}, {88, 78}, {89, 79}, {90, 80}};
```
Why are we modifying the existing variable?
Did we not handle this earlier?

fiona-gladwin commented on rocAL/include/meta_data/meta_data.h:
```diff
@@ -267,7 +289,7 @@ class MetaDataBatch {
    protected:
     MetaDataInfoBatch _info_batch;
-    MetaDataType _type;
+    MetaDataType _type = MetaDataType::Label;
```
Why are we setting here?

fiona-gladwin commented on rocAL/include/pipeline/master_graph.h:
```diff
@@ -131,9 +162,21 @@ class MasterGraph {
+                                        std::vector<std::vector<int>> &sel_mask_ids,
+                                        bool reindex_mask);
+    void set_random_mask_pixel_config(bool is_foreground, int value, bool is_threshold);
+    
```
Remove empty line

fiona-gladwin commented on rocAL/include/pipeline/master_graph.h:
```diff
@@ -242,6 +309,10 @@ class MasterGraph {
     // box IoU matcher variables
     bool _is_box_iou_matcher = false;                                             // bool variable to set the box iou matcher
     BoxIouMatcherInfo _iou_matcher_info;
+    int _random_mask_pixel_value = 0;
+    bool _is_random_mask_pixel_threshold = false;
+    bool _is_random_mask_pixel_foreground = false;
+    std::vector<unsigned> output_random_mask_pixel;
```
Add comment for these variables

fiona-gladwin commented on rocAL/include/pipeline/seed_rng.h:
```diff
@@ -0,0 +1,44 @@
+#include <random>
```
Add copyright info

fiona-gladwin commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -280,23 +284,37 @@ RocalMetaData
     rocalGetMaskCount(RocalContext p_context, int* buf) {
-    if (p_context == nullptr)
-        THROW("Invalid rocal context passed to rocalGetMaskCount")
     unsigned size = 0, count = 0;
+    if (p_context == nullptr)
```
Why the position changed for this?

fiona-gladwin commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -305,26 +323,148 @@ RocalTensorList
     rocalGetMaskCoordinates(RocalContext p_context, int* bufcount) {
     if (p_context == nullptr)
-        THROW("Invalid rocal context passed to rocalGetMaskCoordinates")
+        return nullptr;
```
Why this is changed to return?

SundarRajan28 committed:
Fix review comments: deterministic RNG, RLE, build hygiene

SundarRajan28 committed:
Validate COCO RLE counts and sizes

SundarRajan28 committed:
Fix pixelwise COCO RLE + robust COCO file reader

SundarRajan28 committed:
Fix COCO mask parsing and thread iou flag

SundarRajan28 committed:
Harden random_object_bbox cache and pybind output

SundarRajan28 commented:
Follow-up: addressed the two non-blocking items as well.
- random_object_bbox cache key now mixes in (width,height) to reduce collision risk.
- Python getRandomObjectBBox now returns owning arrays (copy of 4 uints) with a short docstring note.

SundarRajan28 committed:
Revert changes

fiona-gladwin commented on rocAL/include/pipeline/master_graph.h:
```diff
@@ -81,6 +82,41 @@ const __m256i avx_pkdMaskB = _mm256_setr_epi32(0x80808002, 0x80808005, 0x8080800
 #endif
 
+// Cache entry for random_object_bbox caching
+struct RandomObjectBBoxCacheEntry {
```
Can we have this struct along with all the other Random object Bbox related functions in a separate file and be included in MasterGraph?

fiona-gladwin commented on rocAL/include/pipeline/seed_rng.h:
```diff
@@ -0,0 +1,45 @@
+private:
+    std::vector<T> _rngs;
+    int _batch_size;
+};
```
Add a new line after this

fiona-gladwin commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -71,11 +71,15 @@ RocalMetaData
+    if (is_polygon_mask) {
         return context->master_graph->create_coco_meta_data_reader(source_path, is_output, MetaDataReaderType::COCO_META_DATA_READER, MetaDataType::PolygonMask, ltrb, is_box_encoder, avoid_class_remapping, aspect_ratio_grouping, is_box_iou_matcher);
+    } else if (is_pixelwise_mask) {
+        return context->master_graph->create_coco_meta_data_reader(source_path, is_output, MetaDataReaderType::COCO_META_DATA_READER, MetaDataType::PixelwiseMask, ltrb, is_box_encoder, avoid_class_remapping, aspect_ratio_grouping, is_box_iou_matcher);
```
Instead of else if, use a ternary condition to decide the MetadataType and pass that variable to the call, instead of repeating same call twice

fiona-gladwin commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -73,6 +78,8 @@ void COCOMetaDataReader::lookup(const std::vector<std::string> &image_names) {
             _output->get_mask_polygons_count_batch()[i] = it->second->get_polygon_count();
             _output->get_mask_vertices_count_batch()[i] = it->second->get_vertices_count();
         }
+        if (_output->get_metadata_type() == MetaDataType::PixelwiseMask)
```
This should be an else if condition,

fiona-gladwin commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -137,8 +149,203 @@ void COCOMetaDataReader::print_map_contents() {
+            } else {
+                continue;
+            }
+            label_rles[label].push_back(M);
```
Use emplace_back instead of push_back in all instances

fiona-gladwin commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -214,8 +423,14 @@ void COCOMetaDataReader::read_all(const std::string &path) {
-                _label_info.insert(std::make_pair(id, continuous_idx));
-                continuous_idx++;
+                category_ids.push_back(id);
+            }
```
Will this work for the general Metadata reader as well?

fiona-gladwin commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -244,13 +463,48 @@ void COCOMetaDataReader::read_all(const std::string &path) {
-                                parser.SkipArray();
+                            // COCO RLE segmentation is encoded as an object (commonly when iscrowd == 1).
+                            // PolygonMask metadata expects polygon arrays; skip object-valued segmentation to avoid asserts on valid COCO JSON.
+                            if (_output->get_metadata_type() == MetaDataType::PolygonMask) {
```
For Polygon mask will this work? Have we tested it?

fiona-gladwin commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -273,24 +527,58 @@ void COCOMetaDataReader::read_all(const std::string &path) {
+                        if (rle_info.h <= 0) rle_info.h = image_size.h;
+                        if (rle_info.w <= 0) rle_info.w = image_size.w;
+	                        if (rle_info.h != image_size.h || rle_info.w != image_size.w) {
+	                            std::cerr << "WARNING: RLE mask size mismatch for " << itr->second << " (mask "
```
Use WRN statements if applicable

fiona-gladwin commented on rocAL/source/meta_data/coco_meta_data_reader.cpp:
```diff
@@ -312,12 +600,30 @@ void COCOMetaDataReader::read_all(const std::string &path) {
             auto _it_label = _label_info.find(bb_labels[i]);
-            int cnt_idx = _avoid_class_remapping ? _it_label->first : _it_label->second;
-            continuous_label_id.push_back(cnt_idx);
+            if (_it_label == _label_info.end()) {
```
If it is not present should we still push the labels?

fiona-gladwin commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -305,26 +323,148 @@ RocalTensorList
+        size_t meta_data_batch_size = meta_data.second->get_labels_batch().size();
+        if (context->user_batch_size() != meta_data_batch_size)
+            THROW("meta data batch size is wrong " + TOSTR(meta_data_batch_size) + " != " + TOSTR(context->user_batch_size()))
+        auto &polygon_counts_batch = meta_data.second->get_mask_polygons_count_batch();
+        auto &vertices_counts_batch = meta_data.second->get_mask_vertices_count_batch();
+        if (polygon_counts_batch.size() != meta_data_batch_size || vertices_counts_batch.size() != meta_data_batch_size)
+            THROW("Polygon count batch size mismatch")
```
Why don't we directly use context->user_batch_size and not get the labels batch in the line 336, instead we can use directly here

fiona-gladwin commented on rocAL/source/api/rocal_api_meta_data.cpp:
```diff
@@ -305,26 +323,148 @@ RocalTensorList
+RocalTensorList
+    ROCAL_API_CALL
+    rocalSelectMask(RocalContext p_context,
+                    std::vector<int> mask_ids,
```
Accept as reference

fiona-gladwin commented on rocAL/CMakeLists.txt:
```diff
@@ -253,11 +253,14 @@ if(${BUILD_ROCAL})
-                                ${CAFFE_PROTO_SRCS} ${CAFFE2_PROTO_SRCS} ${CAFFE2_PROTO_HEADERS} ${rocdecode_SRCS} ${ROCAL_PROTO_SRCS} ${ROCAL_PROTO_HEADERS})
+    add_library(${PROJECT_NAME} SHARED ${SOURCES} ${TF_PROTO_SRCS} ${TF_PROTO_HEADERS} ${CAFFE_PROTO_HEADERS}
+                                ${CAFFE_PROTO_SRCS} ${CAFFE2_PROTO_SRCS} ${CAFFE2_PROTO_HEADERS} ${rocdecode_SRCS} ${ROCAL_PROTO_SRCS} ${ROCAL_PROTO_HEADERS} ${PYCOCOTOOLS_SOURCES})
+
```
The include of this 3rd party lib needs to be mentioned in the docs right, please check

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +451,477 @@ MasterGraph::reset() {
     return Status::OK;
 }
 
```
All these helper functions instead of keeping in MasterGraph, please organize it into a different file in some other location

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +451,477 @@ MasterGraph::reset() {
+}
+
+void MasterGraph::get_label_boundingboxes(std::vector<std::vector<std::pair<unsigned, unsigned>>> &boxes,
+                                          std::vector<std::pair<unsigned, unsigned>> ranges,
```
All the vectors please pass as references

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +451,477 @@ MasterGraph::reset() {
+        }
+    }
+
+    std::pair<unsigned, unsigned> lo = std::make_pair(0, 0);
+    std::pair<unsigned, unsigned> hi = std::make_pair(0, 0);
+    lo.first = origin[0];
+    hi.first = origin[0] + 1;
+    lo.second = origin[1];
+    hi.second = origin[1] + 1;
```
Directly assign the origin values during declaration itself, reducing 4 lines

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +451,477 @@ MasterGraph::reset() {
+                    _output_random_object_bbox[id][3] = width;
+                    break;
+                case RandomObjectBBoxFormat::OUT_ANCHORSHAPE:
+                    _output_random_object_bbox[id][0] = 0;
```
What is the difference between the above 2 cases, we are doing the same operation?

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +451,477 @@ MasterGraph::reset() {
+
+        // If no foreground labels found, return full image
+        if (unique_labels.empty()) {
+            switch (format) {
```
Check the difference in the case statement definitions

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +451,477 @@ MasterGraph::reset() {
+            // Use cached data
+            nbox = cache_entry->total_boxes[label_selected];
+            if (nbox > 0 && cache_entry->Get(boxes, label_selected)) {
+                // Boxes loaded from cache
```
Why is this empty?

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +451,477 @@ MasterGraph::reset() {
+
+        if (pick_box_id < 0 || pick_box_id >= static_cast<int>(boxes.size()) || boxes[pick_box_id].size() < 2) {
+            // No valid box found, return full image
+            switch (format) {
+                case RandomObjectBBoxFormat::OUT_BOX:
+                case RandomObjectBBoxFormat::OUT_STARTEND:
+                    _output_random_object_bbox[id][0] = 0;
+                    _output_random_object_bbox[id][1] = 0;
+                    _output_random_object_bbox[id][2] = height;
+                    _output_random_object_bbox[id][3] = width;
+                    break;
+                case RandomObjectBBoxFormat::OUT_ANCHORSHAPE:
+                    _output_random_object_bbox[id][0] = 0;
+                    _output_random_object_bbox[id][1] = 0;
+                    _output_random_object_bbox[id][2] = height;
+                    _output_random_object_bbox[id][3] = width;
+                    break;
+                default:
+                    assert(!"Unreachable code");
+            }
```
This is present multiple times, please check if you can have this in a func
Repititive code

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -450,6 +451,477 @@ MasterGraph::reset() {
+        auto &selected_box = boxes[pick_box_id];
+        switch (format) {
+            case RandomObjectBBoxFormat::OUT_BOX:
+                _output_random_object_bbox[id][0] = selected_box[0].first;
```
OUT_BOX, OUT_STARTEND both are same ?

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -1079,11 +1551,30 @@ TensorListVector* MasterGraph::create_coco_meta_data_reader(const char *source_p
+        default_mask_info.set_metadata();
+        _meta_data_buffer_size.emplace_back(_user_batch_size * default_mask_info.data_size());
+        dims = {2};
+        default_random_mask_pixel_info = TensorInfo(dims, _mem_type, RocalTensorDataType::INT32);
+        default_random_mask_pixel_info.set_metadata();
+        _meta_data_buffer_size.emplace_back(_user_batch_size * default_random_mask_pixel_info.data_size());
+        dims = {4};
+        default_random_object_bbox_info = TensorInfo(std::move(dims), _mem_type, RocalTensorDataType::INT32);
+        default_random_object_bbox_info.set_metadata();
```
Will these 2 be present for all the time we get a pixel wise mask or only in special cases?
In that case better to add if else conditions

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -1099,26 +1590,199 @@ TensorListVector* MasterGraph::create_coco_meta_data_reader(const char *source_p
+            if (metadata_type == MetaDataType::PolygonMask) {
+                auto select_mask_polygon_info = default_select_mask_polygon_info;
+                _select_mask_polygon_list.push_back(new Tensor(select_mask_polygon_info, "select_mask_polygon_" + get_tensor_uid()));
+            } else {
+                auto random_mask_pixel_info = default_random_mask_pixel_info;
+                _random_mask_pixel_list.push_back(new Tensor(random_mask_pixel_info, "random_mask_pixel_" + get_tensor_uid()));
+                auto random_object_bbox_info = default_random_object_bbox_info;
+                _random_object_bbox_list.push_back(new Tensor(random_object_bbox_info, "random_object_bbox_" + get_tensor_uid()));
```
Create new tensors only if random mask and random object bbox is required

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -1099,26 +1590,199 @@ TensorListVector* MasterGraph::create_coco_meta_data_reader(const char *source_p
+    return start[idx] + (val - foreground_count[idx]);
+}
+
+TensorList *MasterGraph::get_random_mask_pixel(rocalTensorList *input) {
```
Can be moved to different file

fiona-gladwin commented on rocAL/source/pipeline/master_graph.cpp:
```diff
@@ -1546,24 +2210,91 @@ TensorList *MasterGraph::bbox_meta_data() {
     return &_mask_tensor_list;
 }
 
+TensorList *MasterGraph::get_select_mask_polygon(rocalTensorList *mask_data,
+                                                 std::vector<std::vector<int>> polygon_counts,
+                                                 std::vector<std::vector<std::vector<int>>> vertices_counts,
+                                                 std::vector<int> mask_ids,
+                                                 std::vector<std::vector<int>> &sel_vertices_counts,
+                                                 std::vector<std::vector<int>> &sel_mask_ids,
+                                                 bool reindex_mask) {
```
Please pass only references

SundarRajan28 committed:
Resolve review comments

SundarRajan28 committed:
Update enum usage

SundarRajan28 committed:
Merge remote-tracking branch 'origin' into semantic_segm_support
